fix: async writes and deletes (#4015)
influxdata/docs-v2: content/influxdb/v2.2/write-data/troubleshoot.md

{{% /cloud-only %}}

{{% oss-only %}}

If InfluxDB OSS successfully writes all the request data to the bucket, then it returns success (HTTP `204` status code).
The first rejected point in a batch causes InfluxDB to reject the entire batch and respond with an [HTTP error status](#review-http-status-codes).

{{% /oss-only %}}

### Review HTTP status codes

InfluxDB uses conventional HTTP status codes to indicate the success or failure of a request.
Write requests return the following status codes:

{{% cloud-only %}}

| HTTP response code              | Message                                                                 | Description    |
| :-------------------------------| :---------------------------------------------------------------        | :------------- |
| `204 "Success"`                 |                                                                         | If InfluxDB validated the request data format and queued the data for writing to the bucket |
| `400 "Bad request"`             | `message` contains the first malformed line                             | If data is malformed    |

> | `401 "Unauthorized"`            |                                                                         | If the [`Authorization: Token` header](/influxdb/cloud/api-guide/api_intro/#authentication) is missing or malformed or if the [API token](/influxdb/cloud/api-guide/api_intro/#authentication) doesn't have [permission](/influxdb/cloud/security/tokens/) to write to the bucket |

Author: I copied this pattern from the [cloud limits page](https://docs.influxdata.com/influxdb/cloud/account-management/limits/). Should we remove those?

- Verify the timestamps match the [precision parameter](/influxdb/v2.2/write-data/#timestamp-precision).
- Minimize payload size and network errors by [optimizing writes](/influxdb/v2.2/write-data/best-practices/optimize-writes/).

{{% /oss-only %}}

{{% cloud-only %}}
If you notice data is missing in your bucket, do the following:

- Check the `message` property in the response body for details about the error--for example, `partial write error` indicates [rejected points](#troubleshoot-rejected-points).
- Check for [rejected points](#troubleshoot-rejected-points) in your organization's `_monitoring` bucket.
- Verify all lines contain valid syntax ([line protocol](/influxdb/cloud/reference/syntax/line-protocol/) or [CSV](/influxdb/cloud/reference/syntax/annotated-csv/)). See how to [find parsing errors](#find-parsing-errors).
- Verify the data types match the [series](/influxdb/cloud/reference/key-concepts/data-elements/#series) or [bucket schema](/influxdb/cloud/organizations/buckets/bucket-schema/). See how to resolve [explicit schema rejections](#resolve-explicit-schema-rejections).

> - Verify the timestamps match the [precision parameter](/influxdb/cloud/write-data/#timestamp-precision).

Author: The flow isn't obvious. We expect the user to decide how they're going to write (no-/code). Then, precision use is in those pages.

## Troubleshoot rejected points

{{% oss-only %}}

InfluxDB rejects points for the following reasons:

- The **batch** contains another point with the same series, but one of the fields has a different value type.
- The **bucket** contains another point with the same series, but one of the fields has a different value type.

Check for [field type](/influxdb/v2.2/reference/key-concepts/data-elements/#field-value) differences between the missing data point and other points that have the same [series](/influxdb/v2.2/reference/key-concepts/data-elements/#series)--for example, did you attempt to write `string` data to an `int` field?

{{% /oss-only %}}

{{% cloud-only %}}

InfluxDB may have rejected points even if the HTTP request returned "Success".
InfluxDB logs rejected data points and associated errors to your organization's `_monitoring` bucket.

- [Review rejected points](#review-rejected-points)
  - [Find parsing errors](#find-parsing-errors)
  - [Find data type conflicts and schema rejections](#find-data-type-conflicts-and-schema-rejections)
- [Resolve data type conflicts](#resolve-data-type-conflicts)
- [Resolve explicit schema rejections](#resolve-explicit-schema-rejections)

> ### Review rejected points

Reviewer: Maybe be nice to add example that shows an example query that shows how to find rejected points/.

And/or clarify parsing errors and data type conflicts/schema rejections are a subset of rejected points?

To get a log of rejected points, query the [`rejected_points` measurement](/influxdb/cloud/reference/internals/system-buckets/#_monitoring-bucket-schema) in your organization's `_monitoring` bucket.
To more quickly locate `rejected_points`, keep the following in mind:

> - If your line protocol batch contains single lines with multiple [fields](/influxdb/cloud/reference/syntax/line-protocol/#field-set), InfluxDB logs an entry for each point (each unique field) that is rejected.

Reviewer: Is the limiting factor _both_ single lines and multiple fields in a single line? 

For example, what happens if there's one long line with multiple fields?

Author: > Is the limiting factor _both_ single lines and multiple fields in a single line?
> 
> For example, what happens if there's one long line with multiple fields?

Is this better:

If your line protocol batch contains multiple [fields](/influxdb/cloud/reference/syntax/line-protocol/#field-set) per line, InfluxDB logs an entry for each point (each unique field) that is rejected.

| `measurement` | Measurement name of the point                                                                                                                             |
| `field`       | Name of the field that caused the rejection                                                                                                               |
| `reason`      | Brief description of the problem. See specific reasons in [field type conflicts and schema rejections](#find-field-type-conflicts-and-schema-rejections). |
| `gotType`     | Received [field](/influxdb/cloud/reference/key-concepts/data-elements/#field-value) type: `Boolean`, `Float`, `Integer`, `String`, or `UnsignedInteger`   |
| `wantType`    | Expected [field](/influxdb/cloud/reference/key-concepts/data-elements/#field-value) type: `Boolean`, `Float`, `Integer`, `String`, or `UnsignedInteger`   |
| `<timestamp>` | Time the rejected point was logged                                                                                                                        |

#### Find parsing errors

If InfluxDB can't parse a line (e.g. due to syntax problems), the response `message` might not provide details.

> To find parsing error details, query `rejected_points` entries that contain the `error` field (instead of the `count` field).

Reviewer: ```suggestion
To find parsing error details, query `rejected_points` entries that contain the `error` field.
```

    |> filter(fn: (r) => r._measurement == "rejected_points")
    |> filter(fn: (r) => r._field == "count")
```

### Resolve data type conflicts

When you write to a bucket that has the `implicit` schema type, InfluxDB compares new points to points that have the same [series](/influxdb/cloud/reference/key-concepts/data-elements/#series).
If a point has a field with a different data type than the series, InfluxDB rejects the point and logs a `rejected_points` entry.
The `rejected_points` entry contains one of the following reasons:

| Reason                              | Meaning                                                                                                       |
|:------                              |:-------                                                                                                       |
| `type conflict in batch write`      | The **batch** contains another point with the same series, but one of the fields has a different value type.  |
| `type conflict with existing data`  | The **bucket** contains another point with the same series, but one of the fields has a different value type. |

### Resolve explicit schema rejections

> Buckets with the [`explicit` schema type] use [explicit bucket schemas](/influxdb/cloud/organizations/buckets/bucket-schema/).
> When you write to a bucket that uses explicit bucket schemas,
> InfluxDB rejects data that don't conform to one of the configured schemas.
> 
> Learn how to interpret `rejected_points` logging for [explicit bucket schemas](/influxdb/cloud/organizations/buckets/bucket-schema/).

Reviewer: ```suggestion
If you write to a bucket with an [explicit schema](/influxdb/cloud/organizations/buckets/bucket-schema/), the data must conform to the schema. Otherwise, InfluxDB rejects the data.

Do the following to interpret explicit schema rejections:
```

- [Detect a measurement mismatch](#detect-a-measurement-mismatch)
- [Detect a field type mismatch](#detect-a-field-type-mismatch)

> ##### Detect a measurement mismatch

Reviewer: ```suggestion
#### Detect a measurement mismatch
```

```
airSensors,sensorId=TLM0201 temperature=73.97,humidity=35.23,co=0.48 1637014074
```

The line has an `airSensors` measurement and three fields (`temperature`, `humidity`, and `co`).
If you try to write this data to a bucket that has the [`explicit` schema type](/influxdb/cloud/organizations/buckets/bucket-schema/) and doesn't have an `airSensors` schema, the `/api/v2/write` InfluxDB API returns an error and the following data:

```json
{
  "code": "invalid",
  "message": "3 out of 3 points rejected (check rejected_points in your _monitoring bucket for further information)"
}
```

InfluxDB logs three `rejected_points` entries, one for each field.

| _measurement    | _field | _value | field       | measurement | reason                            |
|:----------------|:-------|:-------|:------------|:------------|:----------------------------------|
| rejected_points | count  | 1      | humidity    | airSensors  | measurement not allowed by schema |
| rejected_points | count  | 1      | co          | airSensors  | measurement not allowed by schema |
| rejected_points | count  | 1      | temperature | airSensors  | measurement not allowed by schema |

> ##### Detect a field type mismatch

Reviewer: ```suggestion
#### Detect a field type mismatch
```

|:------------------------------------|:-----------------------------------------------------------------------------------------------------|
| `field type mismatch with schema`   | The point has the same measurement as a configured schema and they have different field value types. |

Consider a bucket that has the following `airSensors` [`explicit bucket schema`](/influxdb/cloud/organizations/buckets/bucket-schema/):

```json
{
 "name": "airSensors",
 "columns": [
   {
     "name": "time",
     "type": "timestamp"
   },
   {
     "name": "sensorId",
     "type": "tag"
   },
   {
     "name": "temperature",
     "type": "field",
     "dataType": "float"
   },
   {
     "name": "humidity",
     "type": "field",
     "dataType": "float"
   },
   {
     "name": "co",
     "type": "field",
     "dataType": "float"
   }
 ]
}
```

The following [line protocol](/influxdb/cloud/reference/syntax/line-protocol/) data has an `airSensors` measurement, a `sensorId` tag, and three fields (`temperature`, `humidity`, and `co`).

```
airSensors,sensorId=L1 temperature=90.5,humidity=70.0,co=0.2 1637014074
airSensors,sensorId=L1 temperature="90.5",humidity=70.0,co=0.2 1637014074
```

> In the example data, one of the points has a `temperature` field value with the _string_ data type.

Reviewer: ```suggestion
In the example data above, the first point has a `temperature` field value with the _string_ data type.
```

> However, the `airSensors` schema requires `temperature` to have the _float_ data type.
> If you try to write the example data to the `airSensors` bucket schema,
> InfluxDB returns a `400` error and a message that describes the result:

Reviewer: ```suggestion
Because the `airSensors` schema requires `temperature` to have the _float_ data type, 
InfluxDB returns a `400` error and a message that describes the result:
```
